File: leetcode/1-100/28-find-the-index-of-the-first-occurrence-in-a-string/bench/brute_force.c
/*
 * LeetCode 28 — brute-force sliding-window strStr, C mirror.
 * Algorithmic peer of bench/brute_force.{kara,rs,py}. Same N, M, K,
 * adversarial input, same sink.
 */
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>

static int64_t str_str(const uint8_t *haystack, int64_t hn,
                       const uint8_t *needle, int64_t nn) {
    if (nn == 0) {
        return 0;
    }
    if (nn > hn) {
        return -1;
    }
    for (int64_t i = 0; i <= hn - nn; i++) {
        int64_t j = 0;
        while (j < nn && haystack[i + j] == needle[j]) {
            j++;
        }
        if (j == nn) {
            return i;
        }
    }
    return -1;
}

int main(void) {
    const size_t N = 2000000;
    const size_t M = 16;

    uint8_t *haystack = (uint8_t *)malloc(N);
    for (size_t i = 0; i < N; i++) {
        haystack[i] = 'a';
    }
    haystack[N - 1] = 'b';
    uint8_t *needle = (uint8_t *)malloc(M);
    for (size_t i = 0; i < M; i++) {
        needle[i] = 'a';
    }
    needle[M - 1] = 'b';

    int64_t total = 0;
    for (int iter = 0; iter < 10; iter++) {
        total += str_str(haystack, (int64_t)N, needle, (int64_t)M);
    }
    printf("%lld\n", (long long)total);

    free(haystack);
    free(needle);
    return 0;
}

File: leetcode/1-100/28-find-the-index-of-the-first-occurrence-in-a-string/brute_force.py
"""LeetCode #28: Find the Index of the First Occurrence in a String —
brute-force sliding window.

Algorithmic mirror of brute_force.kara. Output format matches line-for-line so
the two can be diffed directly.
"""

from __future__ import annotations


def str_str(haystack: str, needle: str) -> int:
    h = haystack.encode()
    nee = needle.encode()
    hn = len(h)
    nn = len(nee)
    if nn == 0:
        return 0
    if nn > hn:
        return -1
    i = 0
    while i <= hn - nn:
        j = 0
        while j < nn and h[i + j] == nee[j]:
            j += 1
        if j == nn:
            return i
        i += 1
    return -1


def report(haystack: str, needle: str) -> None:
    r = str_str(haystack, needle)
    print(f"{haystack} ~ {needle}: {r}")


def main() -> None:
    report("sadbutsad", "sad")        # expect: 0
    report("leetcode", "leeto")       # expect: -1
    report("hello", "ll")             # expect: 2
    report("a", "a")                  # expect: 0
    report("", "")                    # expect: 0
    report("abc", "")                 # expect: 0
    report("aaa", "aaaa")             # expect: -1
    report("mississippi", "issip")    # expect: 4
    report("abababcabab", "ababc")    # expect: 4
    report("aaaaab", "aab")           # expect: 3
    report("abc", "abc")              # expect: 0
    report("abc", "c")                # expect: 2


if __name__ == "__main__":
    main()
